src/EventStore.Core/Services/VNode/LastIndexableRecordFinder.cs

```diff
@@ -0,0 +1,43 @@
+﻿using EventStore.Core.TransactionLog.Chunks;
+using EventStore.Core.TransactionLog.LogRecords;
+using Serilog;
+
+namespace EventStore.Core.Services.VNode {
+	public class LastIndexableRecordFinder : ILastIndexableRecordFinder {
+		private readonly ILogger _log = Serilog.Log.ForContext<LastIndexableRecordFinder>();
+		private readonly TFChunkDb _db;
+
+		public LastIndexableRecordFinder(TFChunkDb db) {
+			_db = db;
+		}
+
+		public long GetLastIndexableRecordPosition() {
+			var reader = new TFChunkReader(
+				_db,
+				_db.Config.WriterCheckpoint,
+				initialPosition: _db.Config.WriterCheckpoint.Read(),
+				optimizeReadSideCache: _db.Config.OptimizeReadSideCache);
+
+			while (true) {
+				var result = reader.TryReadPrev();
+
+				if (!result.Success) {
+					_log.Information("No indexable records in database");
+					return -1;
+				}
+
+				// coupled to the types of log record that the index indexes
+				// make this more explicit in future versions
+				if (result.LogRecord.RecordType == LogRecordType.Commit) {
+					_log.Information("Last indexable record is a commit at {position}", result.LogRecord.LogPosition);
+					return result.LogRecord.LogPosition;
+				}
+
+				if (result.LogRecord.RecordType == LogRecordType.Prepare) {
```

Comment:
@timothycoleman should the case where a prepare is part of a transaction that has not yet been committed be taken into consideration here? (since it should not be indexable as long as the transaction has not been committed)

Reply:
🤦‍♂️ yes definitely! good catch

Reply:
taken a different approach: whenever we write an epoch we immediately write an $epoch-information event to the $epoch-information stream. This allows the InaugurationManager to know where the
index checkpoint is supposed to reach. As a bonus this also gives the users a way to get details about the epochs. These events will not be present in an unfiltered $all read, for backwards
compatibilty, but can be read directly from their stream.

Reply:
nice idea, thanks!